### PR TITLE
RET-3820 + RET-4081 - Send case linking emails 

### PR DESCRIPTION
**Pull request by marcjudge**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RET-3820
https://tools.hmcts.net/jira/browse/RET-4081

### Change description ###
Send case linking emails

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by brendanmchugh — APPROVED** [review]

**Review by JackReevies — CHANGES_REQUESTED** [review]
Might need to look into fast forwarding master changes into this, we don't use application.properties anymore for example.

I'm also not seeing any apiTests here

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksController.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                newCaseLinks);
+        return ResponseEntity.ok(CCDCallbackResponse.builder()
+                .data(ccdRequest.getCaseDetails().getCaseData())
+                .build());
+    }
+
+    /**
+     * Sends email confirmation of case unlinking.
+     */
+    @PostMapping(value = "maintain/aboutToSubmit", consumes = APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksController.java:79]: It is odd but 'Maintain' its the name of the case linking event so I want to keep it consistent

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksController.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "Sent successfully",
+            content = {
+                @Content(mediaType = "application/json",
+                        schema = @Schema(implementation = CCDCallbackResponse.class))
+            }),
+        @ApiResponse(responseCode = "400", description = "Bad Request"),
+        @ApiResponse(responseCode = "500", description = "Internal Server Error")
+    })
+    public ResponseEntity<CCDCallbackResponse> unLinkCaseSubmitted(
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksController.java]: Renamed

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksController.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "Sent successfully",
+            content = {
+                @Content(mediaType = "application/json",
+                        schema = @Schema(implementation = CCDCallbackResponse.class))
+            }),
+        @ApiResponse(responseCode = "400", description = "Bad Request"),
+        @ApiResponse(responseCode = "500", description = "Internal Server Error")
+    })
+    public ResponseEntity<CCDCallbackResponse> createCaseLinkAbout(
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksController.java]: Updated

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksController.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+        if (!verifyTokenService.verifyTokenSignature(userToken)) {
+            log.error(INVALID_TOKEN, userToken);
+            return ResponseEntity.status(FORBIDDEN.value()).build();
+        }
+        SubmitEvent currentCase = caseRetrievalForCaseWorkerService.caseRetrievalRequest(userToken,
+                ccdRequest.getCaseDetails().getCaseTypeId(), ccdRequest.getCaseDetails().getJurisdiction(),
+                ccdRequest.getCaseDetails().getCaseId());
+        ListTypeItem<CaseLink> currentCaseLinks = currentCase.getCaseData().getCaseLinks();
+        ListTypeItem<CaseLink> newCaseLinks = ccdRequest.getCaseDetails().getCaseData().getCaseLinks();
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksController.java]: Me being lazy, thanks for calling it out. Everything moved now

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksController.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksController.java]: Moved

File: charts/et-cos/values.yaml [hunk 6]
```diff
@@ -82,4 +82,6 @@ java:
     CITIZEN_CASE_DETAILS_URL: "https://et-sya.{{ .Values.global.environment }}.platform.hmcts.net/citizen-hub/"
     TSE_NEW_APPLICATION_ADMIN_TEMPLATE_ID: bd7addc3-d1ea-47cf-9249-208a7033d454
     SEND_NOTIFICATION_TEMPLATE_ID: 1b8299d7-5051-4b78-a870-726b6384d8a8
-    RESPOND_NOTIFICATION_NO_RESPONSE_TEMPLATE_ID: d2caad87-2bab-4bc0-b598-01f1a7341f9e
\ No newline at end of file
+    RESPOND_NOTIFICATION_NO_RESPONSE_TEMPLATE_ID: d2caad87-2bab-4bc0-b598-01f1a7341f9e
+    CASE_LINKED_TEMPLATE_ID: 7e8c6fc1-520a-41cd-a85c-9d3e0b22b282;
+    CASE_UNLINKED_TEMPLATE_ID: 8099d7b8-1703-40c3-95e6-464a3ac776cd;
```
[the remark this answers is not in the document]
Comment (marcjudge) [reply charts/et-cos/values.yaml]: Done

File: src/main/resources/application.yaml [hunk 7]
```diff
@@ -164,4 +164,9 @@ feature-flags:
   # noinspection SpringBootApplicationYaml
   hearings-api: ${HEARINGS_API_FLAG:true} # will be used in future releases, remove noinspection once in use
 
+case-links:
+  linked: ${CASE_LINKED_TEMPLATE_ID:7e8c6fc1-520a-41cd-a85c-9d3e0b22b282}
+  unlinked: ${CASE_UNLINKED_TEMPLATE_ID:8099d7b8-1703-40c3-95e6-464a3ac776cd}
```
Comment (JackReevies) [inline src/main/resources/application.yaml:169]: We have a template section further up in this file, is it worth putting these under that section? ie `template.case-links.linked`

File: src/test/integration/resources/application.yaml [hunk 8]
```diff
@@ -111,5 +111,9 @@ case-details-url:
   citizen: dummyUrl
   exui: dummyUrl
 
+case-links:
+  linked: ${CASE_LINKED_TEMPLATE_ID:7e8c6fc1-520a-41cd-a85c-9d3e0b22b282}
+  unlinked: ${CASE_UNLINKED_TEMPLATE_ID:8099d7b8-1703-40c3-95e6-464a3ac776cd}
```
Comment (JackReevies) [inline src/test/integration/resources/application.yaml:116]: Same for this

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                .build();
+
+        SubmitEvent submitEvent = new SubmitEvent();
+        submitEvent.setCaseData(caseDetails.getCaseData());
+
+        when(verifyTokenService.verifyTokenSignature(AUTH_TOKEN)).thenReturn(true);
+    }
+
+    @ParameterizedTest
+    @MethodSource("requests")
```
Comment (JackReevies) [inline src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksControllerTest.java:73]: This is cool, makes our controller tests file shorter. I'm not sure where we sit with these tests now though as @furnivall  disagreed with apiTests for these controllers, is it worth testing that methods are called and/or case data state *after* running the callback (obvs controller dependant here)

**Review by JackReevies — CHANGES_REQUESTED** [review]

File: src/test/integration/resources/application.yaml (hunk 8 above)
Comment (marcjudge) [reply src/test/integration/resources/application.yaml:116]: Done

File: src/main/resources/application.yaml (hunk 7 above)
Comment (marcjudge) [reply src/main/resources/application.yaml]: Done

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksControllerTest.java (hunk 9 above)
Comment (marcjudge) [reply src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/controllers/CaseLinksControllerTest.java:69]: @JackReevies I see your point. For me the main thing this is testing is the request. Should anyone modify the controller(url or body) I'd want these tests to start failing.
For this specific Controller test I don't want to test case data as its not something we are modifying and service returns void

**Review by JackReevies — APPROVED** [review]

**Pull request merged**